### DISCUSSION TITLE
Rotate without moving.

### DISCUSSION THREAD

**Discussion by Tylertron1998** [SixLabors/ImageSharp#1270] [Q&A] [answered]
Hi all! I have a 410x410 image (lets call it the wheel) that I want to overlay onto another 500x800 image (lets call this the background). I want the wheel to be at 250,250 on the background, and to be able to rotate it freely. The issue is when rotating, it seems the bounding box gets resized - causing what is considered to be the 0, 0 point of the wheel to be moved. Is there any way around this? For example, if I rotate the wheel 45 degrees, the bounding box gets bigger as what would be the top left corner moves to where the 'centre' would be, right at the top - pushing the height of the image up, and the same with the width.

**Comment by Tylertron1998** [comment]
this is my current workaround; which feels very hacky.

```csharp
            wheel.Mutate(context =>
            {
                context.Rotate(22.5f);
                var size = context.GetCurrentSize();
                var middleX = size.Width / 2;
                var middleY = size.Height / 2;
                context.Crop(new Rectangle(middleX - 205, middleY - 205, 410, 410));
            });
```

**Comment by JimBobSquarePants** [comment, marked as answer]
> The issue is when rotating, it seems the bounding box gets resized - causing what is considered to be the 0, 0 point of the wheel to be moved

Well yes, that's maths. Unless you want to clip your input your dimensions will be equal to the hypotenuse of the two sides. 

Most rotations want to preserve the original input without clipping so we've built our rotation with that behavior since, as you've demonstrated it's fairly simple to crop the image following a rotation.

**Reply by Tylertron1998** [reply]
Is there a nicer way of doing this? Perhaps with the Transform method? I'm not big with maths so I'm not sure exactly where I'd go from here.

**Reply by antonfirsov** [reply]
You probably need the following chain of transforms. It could be defined by a single matrix, and executed in one step:
1. move the wheel's origo to `(0,0)` 
2. rotate 
3. apply the inverse of step 1

You can use `AffineTransformBuilder` to create a matrix, or just multiply matrices by hand, then use one of the `Transform` [extension methods](https://github.com/SixLabors/ImageSharp/blob/78a584e8482b052d7a9885682299e2f37518d83d/src/ImageSharp/Processing/Extensions/Transforms/TransformExtensions.cs).

**Reply by antonfirsov** [reply]
My previous comment has several edits

**Reply by Tylertron1998** [reply]
when you say "apply the inverse of step 1" - do you mean literal inverse (i.e. (-0, -0)) or back to where it was before?

**Reply by antonfirsov** [reply]
Sorry, my explanation is a bit messy. I meant that if `(cx,cy)` is the center of your wheel, (1) should translate it by `(-cx,-cy)`, while (3) should move it back by `(+cx,+cy)`.

The point is that you can build a single matrix transform combining the 3 steps.

[Here is](https://math.stackexchange.com/a/62248) a better explanation on what I mean, but you don't have to do the calculations manually, `Matrix3x2`, `Matrix3x2Extensions` and `AffineTransformBuilder` all have a bunch of helpful methods to achieve this.

**Reply by Tylertron1998** [reply]
Awesome - I think I get it! Your explanation is great, no worries. Last question; AffineTransformBuilder has Append/Prepend methods. What should I be using; what are the differences?

Thank you for your time and help!

**Reply by JimBobSquarePants** [reply]
Our rotate does that transformation already. It rotates around the center. @Tylertron1998 Can you please post a picture of your expected result because your code sample above suggests you want to retain the same dimensions yet rotate which requires clipping.

**Reply by Tylertron1998** [reply]
@JimBobSquarePants it's a little difficult to show what my expected result is - because I can't actually *generate* that; I think the key bit of information I have missed out here, is that the image that is being rotated is a circle with a completely transparent background - which I wouldn't understand why what is basically 'no data' is being considered when transforming (causing the bounding box to be expanded).

Edit to clarify: I can understand why, generally, an image with a "background" would have this issue, but for an image that is, for all intents and purposes, only a circle, and for it to be rotated around the centre yet have its 'center point' moved is a little unintuitive.

**Reply by Tylertron1998** [reply]
After a little thought, a much easier solution would be the ability to change what is considered 0,0 in an image. Then, I could move 0,0 to the center of the circle, and no matter what rotation is applied, when adding that image atop *another* image, I will know whatever coordinates I choose, the centre of the circle will *always* land there.

**Reply by Tylertron1998** [reply]
After a little more though (sorry for all of the comments here) there is an easier solution.

```csharp
                        wheel.Mutate(context =>
			{
				context.Rotate(-rotationAmount);
			});

			image.Mutate(context =>
			{
                                // I want the center of the wheel to be at exactly (250, 250)
				context.DrawImage(wheel, new Point(250 - (wheel.Width / 2), 250 - (wheel.Height / 2)), 1f);
			});
```

this is definitely a workaround - though, wouldn't work for non square images (I think?) so again, being able to change where (0, 0) is on an image would be a great fix for all other cases.

I still don't feel exactly comfortable with calling this the 'best way' of doing it, though.

**Reply by JimBobSquarePants** [reply]
@Tylertron1998 The bounds *have* to change to accommodate the dimensions of the rotated element.

Think about it like this... Your circle isn't a circle, it's a square canvas upon which some of the pixel data has been filled in to represent a circle, Surrounding pixels are empty, but represent memory that has to exist to house the circle. 

Given a 100x100 pixel circle you have a 100x100 pixel square. If you rotate the square by 45 degrees the height and width of the square become the length of the hypotenuse of the original lengths. Which, when rounded becomes 141px. (I wonder @antonfirsov should we be rounding up here rather than down?)

![image](https://user-images.githubusercontent.com/385879/87404353-f2b63000-c5b5-11ea-92cc-4c73ad338619.png)
 
I'll use a black square to demonstrate. Here's the original.

![black](https://user-images.githubusercontent.com/385879/87404137-a79c1d00-c5b5-11ea-9688-34d0036d20d3.png)

Now I'll rotate it by 45 degrees. The dimensions of the square itself remain unchanged but the canvas has to expand to accommodate the rotated hypotenuse.

![rotate](https://user-images.githubusercontent.com/385879/87404162-b4207580-c5b5-11ea-97c7-9153ec532595.png)

Here's what would happen if we kept the same original dimensions.

![rotate-crop](https://user-images.githubusercontent.com/385879/87404187-bb478380-c5b5-11ea-87bd-6b90f903eba6.png)

I was able to crop the rotated square using the following code. As you can see it's size agnostic and will work for whatever input dimensions. 

``` c#
using (var image = new Image<Rgba32>(100, 100))
{
    Size originalSize = image.Size();
    image.Mutate(
        x =>
        {
            x.BackgroundColor(Color.Black)
             .Rotate(45F);

            Size size = x.GetCurrentSize();
            SizeF halfSize = (size - originalSize) / 2F;
            Point offset = new Point((int)halfSize.Width, (int)halfSize.Height);
            x.Crop(new Rectangle(offset, originalSize));
        });

    image.Save(Path.Combine(outPath, "rotate-crop.png"));
}
```

In your case you it sounds like you don't want to crop but rather you want to calculate how much to change the top/left origin point to allow drawing in the correct location. Again, that can be calculated for any set of dimensions.

``` c#
// Set your origin point.
Vector2 origin = new Point(0, 0);

// .. Do rotation

// Now calculate your new origin.
Size size = x.GetCurrentSize();
SizeF halfSize = (size - originalSize) / 2F;
origin -= new Vector2(halfSize.Width, halfSize.Height);
var point = Point.Round(origin);
```

Does any of this make sense?

**Reply by Tylertron1998** [reply]
This makes a lot of sense; yep. I guess it was just confusing to me due to the lack of "background" pixels - though, I'd assume, adding a special case for that wouldn't exactly result in clean code and would likely be less intuitive for people who know image manipulation a little more in depth.

Thank you so much for all your help!
